File: app/src/main/java/com/testproject/kaera/ringtestapp/controllers/GalleryController.java
package com.testproject.kaera.ringtestapp.controllers;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import com.bluelinelabs.conductor.ControllerChangeHandler;
import com.bluelinelabs.conductor.ControllerChangeType;
import com.bluelinelabs.conductor.RouterTransaction;
import com.bluelinelabs.conductor.changehandler.FadeChangeHandler;
import com.squareup.picasso.Picasso;
import com.testproject.kaera.ringtestapp.R;
import com.testproject.kaera.ringtestapp.RingApplication;
import com.testproject.kaera.ringtestapp.controllers.DialogController.DialogCallback.Result;
import com.testproject.kaera.ringtestapp.controllers.base.BaseController;
import com.testproject.kaera.ringtestapp.service.command.SaveImageCommand;
import com.testproject.kaera.ringtestapp.ui.util.SwipeLayoutProgressSwitcher;
import com.testproject.kaera.ringtestapp.util.BundleBuilder;
import com.testproject.kaera.ringtestapp.util.SimpleTargetCallback;

import java.net.URI;
import java.net.URISyntaxException;

import javax.inject.Inject;

import butterknife.BindView;
import io.techery.janet.ActionPipe;

import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;

public class GalleryController extends BaseController {

    private static final String KEY_URL = "key_url";
    private Bitmap loadedBitmap;

    @Inject ActionPipe<SaveImageCommand> saveImageCommand;

    @BindView(R.id.thumbnail_image) ImageView imgThumb;
    @BindView(R.id.refresh_layout) SwipeRefreshLayout refreshLayout;

    public GalleryController(String thumbUrl) {
        this(new BundleBuilder()
                .putCharSequence(KEY_URL, thumbUrl)
                .build());
        setHasOptionsMenu(true);
    }

    public GalleryController(Bundle args) {
        super(args);
    }

    @Override protected void onAttach(@NonNull View view) {
        super.onAttach(view);
        requestPermissions(new String[]{WRITE_EXTERNAL_STORAGE}, 0);
        bindPipe(saveImageCommand)
                .afterEach(new SwipeLayoutProgressSwitcher<>(refreshLayout))
                .onSuccess(command -> showToast(R.string.success_message_save_image))
                .onFail((command, throwable) -> showToast(R.string.error_message_save_image));
    }

    @Override
    protected View inflateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
        return inflater.inflate(R.layout.controller_gallery, container, false);
    }

    @Override
    public void onViewBound(@NonNull View view) {
        super.onViewBound(view);
        RingApplication.getComponent().inject(this);
        loadImage();
        refreshLayout.setColorSchemeResources(R.color.colorAccent);
        refreshLayout.setOnRefreshListener(this::loadImage);
    }

    private void loadImage() {
        Picasso.get()
                .load(getArgs().getCharSequence(KEY_URL).toString())
                .into(new SimpleTargetCallback() {
                    @Override public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                        refreshLayout.setRefreshing(false);
                        loadedBitmap = bitmap;
                        imgThumb.setImageBitmap(bitmap);
                    }

                    @Override public void onBitmapFailed(Exception e, Drawable errorDrawable) {
                        super.onBitmapFailed(e, errorDrawable);
                        refreshLayout.setRefreshing(false);
                        showToast(R.string.error_message_load_image);
                    }

                    @Override public void onPrepareLoad(Drawable placeHolderDrawable) {
                        super.onPrepareLoad(placeHolderDrawable);
                        refreshLayout.setRefreshing(true);
                    }
                });
    }

    private void showToast(@StringRes int message) {
        Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onChangeEnded(@NonNull ControllerChangeHandler changeHandler, @NonNull ControllerChangeType changeType) {
        super.onChangeEnded(changeHandler, changeType);
        if (changeType == ControllerChangeType.POP_ENTER && loadedBitmap == null) {
            loadImage();
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.save, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.save) {
            showDialog();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void showDialog() {
        getRouter().pushController(RouterTransaction.with(new DialogController("Conductor", "Do you want to save image")
                .setDialogCallback(this::onResult))
                .popChangeHandler(new FadeChangeHandler())
                .pushChangeHandler(new FadeChangeHandler(false)));
    }

    private void onResult(Result result) {
        if (result == Result.OK) {
            saveImageCommand.send(new SaveImageCommand(loadedBitmap, getImageName(getArgs().getCharSequence(KEY_URL).toString())));
        }
    }

    private String getImageName(String url) {
        String result = "temporary_name";
        try {
            result = new URI(url).getPath().replaceAll("/", "");
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        return result;
    }
}

File: app/src/main/java/com/testproject/kaera/ringtestapp/controllers/GalleryController.kt
package com.testproject.kaera.ringtestapp.controllers

import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.annotation.StringRes
import android.util.Log
import android.util.TypedValue
import android.view.*
import android.widget.Toast
import com.bluelinelabs.conductor.ControllerChangeHandler
import com.bluelinelabs.conductor.ControllerChangeType
import com.bluelinelabs.conductor.RouterTransaction
import com.bluelinelabs.conductor.changehandler.FadeChangeHandler
import com.squareup.picasso.Picasso
import com.testproject.kaera.ringtestapp.R
import com.testproject.kaera.ringtestapp.controllers.base.BaseController
import com.testproject.kaera.ringtestapp.service.command.SaveImageCommand
import com.testproject.kaera.ringtestapp.ui.util.SwipeLayoutProgressSwitcher
import com.testproject.kaera.ringtestapp.util.BundleBuilder
import com.testproject.kaera.ringtestapp.util.SimpleTargetCallback
import com.testproject.kaera.ringtestapp.util.getComponent
import io.techery.janet.ActionPipe
import kotlinx.android.synthetic.main.controller_gallery.view.*
import java.lang.Exception
import java.net.URI
import java.net.URISyntaxException
import javax.inject.Inject

class GalleryController(args: Bundle?) : BaseController(args) {

    companion object {
        var KEY_URL: String = "key_url"
    }

    private var loadedBitmap: Bitmap? = null

    @Inject
    lateinit var saveImageCommand: ActionPipe<SaveImageCommand>

    constructor(thumbUrl: String) : this(BundleBuilder()
            .putCharSequence(KEY_URL, thumbUrl)
            .build()) {
        setHasOptionsMenu(false)
    }

    override fun onAttach(view: View) {
        super.onAttach(view)
        requestPermissions(arrayOf(WRITE_EXTERNAL_STORAGE), 0)
        bindPipe(saveImageCommand)
                .afterEach(SwipeLayoutProgressSwitcher(view.refresh_layout))
                .onSuccess({ showToast(R.string.success_message_save_image) })
                .onFail { t1, t2 -> showToast(R.string.error_message_save_image) }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup): View {
        val view = inflater.inflate(R.layout.controller_gallery, container, false)
        applicationContext!!.getComponent().inject(this)
        loadImage()
        view.refresh_layout.setColorSchemeResources(R.color.colorAccent)
        view.refresh_layout.setOnRefreshListener(this::loadImage)
        return view
    }

    private fun loadImage() {
        Picasso.get()
                .load(args.getCharSequence(KEY_URL)?.toString())
                .resize((480 * resources?.displayMetrics?.density!!).toInt(),
                        (640 * resources?.displayMetrics?.density!!).toInt())
                .centerCrop()
                .into(object : SimpleTargetCallback() {
                    override fun onBitmapFailed(e: Exception?, errorDrawable: Drawable?) {
                        view?.refresh_layout?.isRefreshing = false
                        showToast(R.string.error_message_load_image)
                    }

                    override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
                        view?.refresh_layout?.isRefreshing = false
                        loadedBitmap = bitmap
                        view?.thumbnail_image?.setImageBitmap(bitmap)
                    }
                })
    }

    private fun showToast(@StringRes message: Int) {
        Toast.makeText(applicationContext, message, Toast.LENGTH_SHORT).show()
    }

    override fun onChangeEnded(changeHandler: ControllerChangeHandler, changeType: ControllerChangeType) {
        super.onChangeEnded(changeHandler, changeType)
        if (changeType == ControllerChangeType.POP_ENTER && loadedBitmap == null) {
            loadImage()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.save, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.save) {
            showDialog()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showDialog() {
        router.pushController(RouterTransaction.with(DialogController("Conductor", "Do you want to save image")
                .setDialogCallback({ onResult(it) }))
                .popChangeHandler(FadeChangeHandler())
                .pushChangeHandler(FadeChangeHandler(false)))
    }

    private fun onResult(result: DialogController.DialogCallback.Result) {
        if (result == DialogController.DialogCallback.Result.OK) {
            val title = getImageName(args.getCharSequence(KEY_URL).toString())
            saveImageCommand.send(SaveImageCommand(loadedBitmap, title))
        }
    }

    private fun getImageName(url: String): String {
        var result = "temporary_name"
        try {
            result = URI(url).path.replace("/".toRegex(), "")
        } catch (e: URISyntaxException) {
            e.printStackTrace()
        }

        return result
    }
}